Invalid Alethe proof

**Describe the bug**

On this example, cvc5 produces an incorrect Alethe proof:
```
(set-logic QF_UF)
(declare-const x2 Bool)
(declare-const x Bool)
(declare-sort I 0)
(declare-fun u () I)
(declare-fun o (I I) I)
(declare-fun e () I)
(assert (or x x2))
(assert (or (= u (o u u)) (and (= u (o u u)) (or (not (= (o u (o u u)) (o (o (o e e) (o u (o u u))) (o e e)))) (= (o u (o u u)) (o (o e e) (o (o e e) (o u (o u u)))))))))
(assert (= (o e e) (o u u)))
(assert (or (= (o u (o u u)) (o (o e e) (o (o e e) (o u (o u u))))) (not (= (o u (o u u)) (o (o (o e e) (o u (o u u))) (o e e))))))
(assert (not (= (o u (o u u)) (o (o u u) (o (o e e) (o u (o u u)))))))
(check-sat)
```

This happens with commit 6a0abb70032db7bd747724477ca63372e7c801eb from main. 

To get the proof:
```
> cvc5 --dump-proofs --proof-format-mode=alethe --dag-thresh=0 --proof-granularity=theory-rewrite test.smt2
```
Then `carcara` complains:
```
> carcara check test.prf test.smt2
[ERROR] checking failed on step 't22' with rule 'resolution': pivot was not eliminated: '(not (= (o u (o u u)) (o (o (o e e) (o u (o u u))) (o e e))))'
invalid
```
Checking by hand confirms that carcara is right.

--- a/src/proof/alethe/alethe_post_processor.cpp
+++ b/src/proof/alethe/alethe_post_processor.cpp
@@ -1829,6 +1829,9 @@ bool AletheProofPostprocessCallback::update(Node res,
 bool AletheProofPostprocessCallback::maybeReplacePremiseProof(Node premise,
                                                               CDProof* cdp)
 {
+  // This method is called only when the premise is used as a singleton
+  // clause. If its proof however concludes a non-singleton clause it'll fail
+  // the test below and we must change its proof.
   std::shared_ptr<ProofNode> premisePf = cdp->getProofFor(premise);
   Node premisePfConclusion = premisePf->getArguments()[2];
   if (premisePfConclusion.getNumChildren() <= 2
@@ -1836,6 +1839,25 @@ bool AletheProofPostprocessCallback::maybeReplacePremiseProof(Node premise,
   {
     return false;
   }
+  // If this resolution child is used as a singleton OR but the rule
+  // justifying it concludes a clause, then we are necessarily in this
+  // scenario:
+  //
+  // (or t1 ... tn)
+  // -------------- OR
+  // (cl t1 ... tn)
+  // ---------------- FACTORING/REORDERING
+  // (cl t1' ... tn')
+  //
+  // where what is used in the resolution is actually (or t1' ... tn').
+  //
+  // This happens when (or t1' ... tn') has been added to the SAT solver as
+  // a literal as well, and its node clashed with the conclusion of the
+  // FACTORING/REORDERING step.
+  //
+  // The solution is to *not* use FACTORING/REORDERING (which in Alethe
+  // operate on clauses) but generate a proof to obtain (via rewriting) the
+  // expected node (or t1' ... tn') from the original node (or t1 ... tn).
   NodeManager* nm = NodeManager::currentNM();
   Trace("alethe-proof") << "\n";
   CVC5_UNUSED AletheRule premisePfRule =
@@ -1849,10 +1871,14 @@ bool AletheProofPostprocessCallback::maybeReplacePremiseProof(Node premise,
   std::shared_ptr<ProofNode> premiseChildPf =
       premisePf->getChildren()[0]->getChildren()[0];
   Node premiseChildConclusion = premiseChildPf->getResult();
-  // Note that we need to add this proof node explicitly to cdp
-  // because cdp does not have a step for it. Rather it is only
-  // present in cdp as a descendant of premisePf (which is in cdp), so
-  // if premisePf is to be lost, then so will premiseChildPf.
+  // Note that we need to add this proof node explicitly (i.e., as an ASSUME
+  // step) to cdp because cdp does not have a step for
+  // premiseChildConclusion. Rather it is only present in cdp as a descendant of
+  // premisePf (which is in cdp), so if premisePf is to be lost, then so will
+  // premiseChildPf. By adding the ASSUME step for premiseChildConclusion, a
+  // step will be present in cdp connecting premiseChildConclusion to
+  // premiseChildPf (since by default adding an ASSUME step will not rewrite an
+  // existing proof for a node).
   addAletheStep(AletheRule::ASSUME,
                 premiseChildConclusion,
                 premiseChildConclusion,
@@ -1971,25 +1997,8 @@ bool AletheProofPostprocessCallback::updatePost(
                                 << " / " << newConclusion << std::endl;
         }
       }
-      // If this resolution child is used as a singleton OR but the rule
-      // justifying it concludes a clause, the we are necessarily in this
-      // scenario:
-      //
-      // (or t1 ... tn)
-      // -------------- OR
-      // (cl t1 ... tn)
-      // ---------------- FACTORING/REORDERING
-      // (cl t1' ... tn')
-      //
-      // where what is used in the resolution is actually (or t1' ... tn').
-      //
-      // This happens when (or t1' ... tn') has been added to the SAT solver as
-      // a literal as well, and its node clashed with the conclusion of the
-      // FACTORING/REORDERING step.
-      //
-      // The solution is to *not* use FACTORING/REORDERING (which in Alethe
-      // operate on clauses) but generate a proof to obtain (via rewriting) the
-      // expected node (or t1' ... tn') from the original node (or t1 ... tn).
+      // The premise is used a singleton clause. We must guarantee that its
+      // proof indeed concludes a singleton clause.
       else if (children[0].getKind() == kind::OR)
       {
         Assert(args[polIdx] == d_true && args[pivIdx] == children[0]);

--- a/src/proof/alethe/alethe_post_processor.cpp
+++ b/src/proof/alethe/alethe_post_processor.cpp
@@ -1831,7 +1831,6 @@ bool AletheProofPostprocessCallback::maybeReplacePremiseProof(Node premise,
 {
   std::shared_ptr<ProofNode> premisePf = cdp->getProofFor(premise);
   Node premisePfConclusion = premisePf->getArguments()[2];
-  AletheRule premisePfRule = getAletheRule(premisePf->getArguments()[0]);
   if (premisePfConclusion.getNumChildren() <= 2
       || premisePfConclusion[0] != d_cl)
   {
@@ -1839,10 +1838,13 @@ bool AletheProofPostprocessCallback::maybeReplacePremiseProof(Node premise,
   }
   NodeManager* nm = NodeManager::currentNM();
   Trace("alethe-proof") << "\n";
+  CVC5_UNUSED AletheRule premisePfRule =
+      getAletheRule(premisePf->getArguments()[0]);
+  CVC5_UNUSED AletheRule premiseChildPfRule =
+      getAletheRule(premisePf->getChildren()[0]->getArguments()[0]);
   Assert((premisePfRule == AletheRule::CONTRACTION
           || premisePfRule == AletheRule::REORDERING)
-         && getAletheRule(premisePf->getChildren()[0]->getArguments()[0])
-                == AletheRule::OR);
+         && premiseChildPfRule == AletheRule::OR);
   // get great grand child
   std::shared_ptr<ProofNode> premiseChildPf =
       premisePf->getChildren()[0]->getChildren()[0];

--- a/src/proof/alethe/alethe_post_processor.cpp
+++ b/src/proof/alethe/alethe_post_processor.cpp
@@ -83,7 +83,7 @@ bool AletheProofPostprocessCallback::update(Node res,
                                             CDProof* cdp,
                                             bool& continueUpdate)
 {
-  Trace("alethe-proof") << "- Alethe post process callback " << res << " " << id
+  Trace("alethe-proof") << "...Alethe pre-update " << res << " " << id
                         << " " << children << " / " << args << std::endl;
 
   NodeManager* nm = NodeManager::currentNM();
@@ -1841,7 +1841,7 @@ bool AletheProofPostprocessCallback::updatePost(
 {
   NodeManager* nm = NodeManager::currentNM();
   AletheRule rule = getAletheRule(args[0]);
-  Trace("alethe-proof") << "... finalizer for rule " << rule << " / " << res
+  Trace("alethe-proof") << "...Alethe post-update " << rule << " / " << res
                         << " / args: " << args << std::endl;
   switch (rule)
   {
@@ -1871,7 +1871,7 @@ bool AletheProofPostprocessCallback::updatePost(
       // to its pivot L_1. Since it's the first clause in the resolution it can
       // only be equal to the pivot in the case the polarity is true.
       if (children[0].getKind() == kind::OR
-          && (args[polIdx] != d_true || children[0] != args[pivIdx]))
+          && (args[polIdx] != d_true || args[pivIdx] != children[0]))
       {
         std::shared_ptr<ProofNode> childPf = cdp->getProofFor(children[0]);
         Node childConclusion = childPf->getArguments()[2];
@@ -1905,9 +1905,80 @@ bool AletheProofPostprocessCallback::updatePost(
                         {},
                         *cdp);
           newChildren[0] = newConclusion;
+          Trace("alethe-proof") << "Added OR step for " << childConclusion
+                                << " / " << newConclusion << std::endl;
+        }
+      }
+      // If child is used as a singleton but the premise is a clause, then
+      // inadvertently the result of a FACTORING or REORDERING step over a
+      // clause, obtained from an OR step over a Node, yielded the same Node
+      // of a literal. In this cause we must not use FACTORING or REORDERING
+      // (which in Alethe operate on clauses) but rewrite the original Node
+      // converted into a clause and processed, thus obtained the Node of the
+      // literal being used in the premise.
+      else if (children[0].getKind() == kind::OR)
+      {
+        Assert(args[polIdx] == d_true && args[pivIdx] == children[0]);
+        std::shared_ptr<ProofNode> childPf = cdp->getProofFor(children[0]);
+        Node childConclusion = childPf->getArguments()[2];
+        AletheRule childRule = getAletheRule(childPf->getArguments()[0]);
+        if (childConclusion.getNumChildren() > 2 && childConclusion[0] == d_cl)
+        {
+          hasUpdated = true;
+          Trace("alethe-proof") << "\n";
+          Assert((childRule == AletheRule::CONTRACTION
+                  || childRule == AletheRule::REORDERING)
+                 && getAletheRule(childPf->getChildren()[0]->getArguments()[0])
+                        == AletheRule::OR);
+          // get great grand child
+          std::shared_ptr<ProofNode> greatGrandChildPf =
+              childPf->getChildren()[0]->getChildren()[0];
+          Node greatGrandChildConclusion = greatGrandChildPf->getResult();
+          // Note that we need to add this proof node explicitly to cdp
+          // because cdp does not have a step for it. Rather it is only
+          // present in cdp as a descendant of childPf (which is in cdp), so
+          // if childPf is to be lost, then so will greatGrandChildPf.
+          addAletheStep(AletheRule::ASSUME,
+                        greatGrandChildConclusion,
+                        greatGrandChildConclusion,
+                        {},
+                        {},
+                        *cdp);
+          // equate it to what we expect, use equiv elim and resolution to
+          // obtain a proof the expected
+          Node expectedConclusion = childPf->getResult();
+          Assert(expectedConclusion == children[0]);
+          Node equiv = greatGrandChildConclusion.eqNode(expectedConclusion);
+          addAletheStep(AletheRule::ALL_SIMPLIFY,
+                        equiv,
+                        nm->mkNode(kind::SEXPR, d_cl, equiv),
+                        {},
+                        {},
+                        *cdp);
+          Node equivElim = nm->mkNode(kind::SEXPR,
+                                      {d_cl,
+                                       equiv.notNode(),
+                                       greatGrandChildConclusion.notNode(),
+                                       expectedConclusion});
+          addAletheStep(
+              AletheRule::EQUIV_POS2, equivElim, equivElim, {}, {}, *cdp);
+          Node newChild = nm->mkNode(kind::SEXPR, d_cl, expectedConclusion);
+          addAletheStep(AletheRule::RESOLUTION,
+                        newChild,
+                        newChild,
+                        {equivElim, equiv, greatGrandChildConclusion},
+                        d_resPivots
+                            ? std::vector<Node>{equiv,
+                                                d_false,
+                                                greatGrandChildConclusion,
+                                                d_false}
+                            : std::vector<Node>(),
+                        *cdp);
+          newChildren[0] = newChild;
           Trace("alethe-proof")
-              << "Added OR step in finalizer " << childConclusion << " / "
-              << newConclusion << std::endl;
+              << "Reverted handling as a clause for converting "
+              << greatGrandChildConclusion << " into " << expectedConclusion
+              << std::endl;
         }
       }
       // For all other children C_i the procedure is similar. There is however a
@@ -1963,9 +2034,77 @@ bool AletheProofPostprocessCallback::updatePost(
                           {},
                           *cdp);
             newChildren[i] = conclusion;
+            Trace("alethe-proof") << "Added OR step for " << childConclusion
+                                  << " / " << conclusion << std::endl;
+          }
+        }
+        // As for the first premise, we need to handle the case in which the
+        // premise is a singleton but the rule concluding it yields a clause.
+        else if (children[i].getKind() == kind::OR)
+        {
+          Assert(args[polIdx] == d_false && args[pivIdx] == children[i]);
+          std::shared_ptr<ProofNode> childPf = cdp->getProofFor(children[i]);
+          Node childConclusion = childPf->getArguments()[2];
+          if (childConclusion.getNumChildren() > 2
+              && childConclusion[0] == d_cl)
+          {
+            hasUpdated = true;
+            Trace("alethe-proof") << "\n";
+            AletheRule childRule = getAletheRule(childPf->getArguments()[0]);
+            Assert(
+                (childRule == AletheRule::CONTRACTION
+                 || childRule == AletheRule::REORDERING)
+                && getAletheRule(childPf->getChildren()[0]->getArguments()[0])
+                       == AletheRule::OR);
+            // get great grand child
+            std::shared_ptr<ProofNode> greatGrandChildPf =
+                childPf->getChildren()[0]->getChildren()[0];
+            Node greatGrandChildConclusion = greatGrandChildPf->getResult();
+            // Note that we need to add this proof node explicitly to cdp
+            // because cdp does not have a step for it. Rather it is only
+            // present in cdp as a descendant of childPf (which is in cdp), so
+            // if childPf is to be lost, then so will greatGrandChildPf.
+            addAletheStep(AletheRule::ASSUME,
+                          greatGrandChildConclusion,
+                          greatGrandChildConclusion,
+                          {},
+                          {},
+                          *cdp);
+            // equate it to what we expect, use equiv elim and resolution to
+            // obtain a proof the expected
+            Node expectedConclusion = childPf->getResult();
+            Assert(expectedConclusion == children[i]);
+            Node equiv = greatGrandChildConclusion.eqNode(expectedConclusion);
+            addAletheStep(AletheRule::ALL_SIMPLIFY,
+                          equiv,
+                          nm->mkNode(kind::SEXPR, d_cl, equiv),
+                          {},
+                          {},
+                          *cdp);
+            Node equivElim = nm->mkNode(kind::SEXPR,
+                                        {d_cl,
+                                         equiv.notNode(),
+                                         greatGrandChildConclusion.notNode(),
+                                         expectedConclusion});
+            addAletheStep(
+                AletheRule::EQUIV_POS2, equivElim, equivElim, {}, {}, *cdp);
+            Node newChild = nm->mkNode(kind::SEXPR, d_cl, expectedConclusion);
+            addAletheStep(AletheRule::RESOLUTION,
+                          newChild,
+                          newChild,
+                          {equivElim, equiv, greatGrandChildConclusion},
+                          d_resPivots
+                              ? std::vector<Node>{equiv,
+                                                  d_false,
+                                                  greatGrandChildConclusion,
+                                                  d_false}
+                              : std::vector<Node>(),
+                          *cdp);
+            newChildren[i] = newChild;
             Trace("alethe-proof")
-                << "Added OR step in finalizer " << childConclusion << " / "
-                << conclusion << std::endl;
+                << "Reverted handling as a clause for converting "
+                << greatGrandChildConclusion << " into " << expectedConclusion
+                << std::endl;
           }
         }
       }

--- a/src/proof/alethe/alethe_post_processor.cpp
+++ b/src/proof/alethe/alethe_post_processor.cpp
@@ -1826,6 +1826,66 @@ bool AletheProofPostprocessCallback::update(Node res,
   }
 }
 
+bool AletheProofPostprocessCallback::maybeReplacePremiseProof(Node premise,
+                                                              CDProof* cdp)
+{
+  std::shared_ptr<ProofNode> premisePf = cdp->getProofFor(premise);
+  Node premisePfConclusion = premisePf->getArguments()[2];
+  AletheRule premisePfRule = getAletheRule(premisePf->getArguments()[0]);
+  if (premisePfConclusion.getNumChildren() <= 2
+      || premisePfConclusion[0] != d_cl)
+  {
+    return false;
+  }
+  NodeManager* nm = NodeManager::currentNM();
+  Trace("alethe-proof") << "\n";
+  Assert((premisePfRule == AletheRule::CONTRACTION
+          || premisePfRule == AletheRule::REORDERING)
+         && getAletheRule(premisePf->getChildren()[0]->getArguments()[0])
+                == AletheRule::OR);
+  // get great grand child
+  std::shared_ptr<ProofNode> premiseChildPf =
+      premisePf->getChildren()[0]->getChildren()[0];
+  Node premiseChildConclusion = premiseChildPf->getResult();
+  // Note that we need to add this proof node explicitly to cdp
+  // because cdp does not have a step for it. Rather it is only
+  // present in cdp as a descendant of premisePf (which is in cdp), so
+  // if premisePf is to be lost, then so will premiseChildPf.
+  addAletheStep(AletheRule::ASSUME,
+                premiseChildConclusion,
+                premiseChildConclusion,
+                {},
+                {},
+                *cdp);
+  // equate it to what we expect, use equiv elim and resolution to
+  // obtain a proof the expected
+  Node equiv = premiseChildConclusion.eqNode(premise);
+  addAletheStep(AletheRule::ALL_SIMPLIFY,
+                equiv,
+                nm->mkNode(kind::SEXPR, d_cl, equiv),
+                {},
+                {},
+                *cdp);
+  Node equivElim = nm->mkNode(
+      kind::SEXPR,
+      {d_cl, equiv.notNode(), premiseChildConclusion.notNode(), premise});
+  addAletheStep(AletheRule::EQUIV_POS2, equivElim, equivElim, {}, {}, *cdp);
+  Node newPremise = nm->mkNode(kind::SEXPR, d_cl, premise);
+  addAletheStep(
+      AletheRule::RESOLUTION,
+      newPremise,
+      newPremise,
+      {equivElim, equiv, premiseChildConclusion},
+      d_resPivots
+          ? std::vector<Node>{equiv, d_false, premiseChildConclusion, d_false}
+          : std::vector<Node>(),
+      *cdp);
+  Trace("alethe-proof") << "Reverted handling as a clause for converting "
+                        << premiseChildConclusion << " into " << premise
+                        << std::endl;
+  return true;
+}
+
 // Adds an OR rule to the premises of a step if the premise is not a clause and
 // should not be a singleton. Since CONTRACTION and REORDERING always take
 // non-singletons, this function adds an OR step to their premise if it was
@@ -1909,76 +1969,32 @@ bool AletheProofPostprocessCallback::updatePost(
                                 << " / " << newConclusion << std::endl;
         }
       }
-      // If child is used as a singleton but the premise is a clause, then
-      // inadvertently the result of a FACTORING or REORDERING step over a
-      // clause, obtained from an OR step over a Node, yielded the same Node
-      // of a literal. In this cause we must not use FACTORING or REORDERING
-      // (which in Alethe operate on clauses) but rewrite the original Node
-      // converted into a clause and processed, thus obtained the Node of the
-      // literal being used in the premise.
+      // If this resolution child is used as a singleton OR but the rule
+      // justifying it concludes a clause, the we are necessarily in this
+      // scenario:
+      //
+      // (or t1 ... tn)
+      // -------------- OR
+      // (cl t1 ... tn)
+      // ---------------- FACTORING/REORDERING
+      // (cl t1' ... tn')
+      //
+      // where what is used in the resolution is actually (or t1' ... tn').
+      //
+      // This happens when (or t1' ... tn') has been added to the SAT solver as
+      // a literal as well, and its node clashed with the conclusion of the
+      // FACTORING/REORDERING step.
+      //
+      // The solution is to *not* use FACTORING/REORDERING (which in Alethe
+      // operate on clauses) but generate a proof to obtain (via rewriting) the
+      // expected node (or t1' ... tn') from the original node (or t1 ... tn).
       else if (children[0].getKind() == kind::OR)
       {
         Assert(args[polIdx] == d_true && args[pivIdx] == children[0]);
-        std::shared_ptr<ProofNode> childPf = cdp->getProofFor(children[0]);
-        Node childConclusion = childPf->getArguments()[2];
-        AletheRule childRule = getAletheRule(childPf->getArguments()[0]);
-        if (childConclusion.getNumChildren() > 2 && childConclusion[0] == d_cl)
+        if (maybeReplacePremiseProof(children[0], cdp))
         {
           hasUpdated = true;
-          Trace("alethe-proof") << "\n";
-          Assert((childRule == AletheRule::CONTRACTION
-                  || childRule == AletheRule::REORDERING)
-                 && getAletheRule(childPf->getChildren()[0]->getArguments()[0])
-                        == AletheRule::OR);
-          // get great grand child
-          std::shared_ptr<ProofNode> greatGrandChildPf =
-              childPf->getChildren()[0]->getChildren()[0];
-          Node greatGrandChildConclusion = greatGrandChildPf->getResult();
-          // Note that we need to add this proof node explicitly to cdp
-          // because cdp does not have a step for it. Rather it is only
-          // present in cdp as a descendant of childPf (which is in cdp), so
-          // if childPf is to be lost, then so will greatGrandChildPf.
-          addAletheStep(AletheRule::ASSUME,
-                        greatGrandChildConclusion,
-                        greatGrandChildConclusion,
-                        {},
-                        {},
-                        *cdp);
-          // equate it to what we expect, use equiv elim and resolution to
-          // obtain a proof the expected
-          Node expectedConclusion = childPf->getResult();
-          Assert(expectedConclusion == children[0]);
-          Node equiv = greatGrandChildConclusion.eqNode(expectedConclusion);
-          addAletheStep(AletheRule::ALL_SIMPLIFY,
-                        equiv,
-                        nm->mkNode(kind::SEXPR, d_cl, equiv),
-                        {},
-                        {},
-                        *cdp);
-          Node equivElim = nm->mkNode(kind::SEXPR,
-                                      {d_cl,
-                                       equiv.notNode(),
-                                       greatGrandChildConclusion.notNode(),
-                                       expectedConclusion});
-          addAletheStep(
-              AletheRule::EQUIV_POS2, equivElim, equivElim, {}, {}, *cdp);
-          Node newChild = nm->mkNode(kind::SEXPR, d_cl, expectedConclusion);
-          addAletheStep(AletheRule::RESOLUTION,
-                        newChild,
-                        newChild,
-                        {equivElim, equiv, greatGrandChildConclusion},
-                        d_resPivots
-                            ? std::vector<Node>{equiv,
-                                                d_false,
-                                                greatGrandChildConclusion,
-                                                d_false}
-                            : std::vector<Node>(),
-                        *cdp);
-          newChildren[0] = newChild;
-          Trace("alethe-proof")
-              << "Reverted handling as a clause for converting "
-              << greatGrandChildConclusion << " into " << expectedConclusion
-              << std::endl;
+          newChildren[0] = nm->mkNode(kind::SEXPR, d_cl, children[0]);
         }
       }
       // For all other children C_i the procedure is similar. There is however a
@@ -2043,68 +2059,10 @@ bool AletheProofPostprocessCallback::updatePost(
         else if (children[i].getKind() == kind::OR)
         {
           Assert(args[polIdx] == d_false && args[pivIdx] == children[i]);
-          std::shared_ptr<ProofNode> childPf = cdp->getProofFor(children[i]);
-          Node childConclusion = childPf->getArguments()[2];
-          if (childConclusion.getNumChildren() > 2
-              && childConclusion[0] == d_cl)
+          if (maybeReplacePremiseProof(children[i], cdp))
           {
             hasUpdated = true;
-            Trace("alethe-proof") << "\n";
-            AletheRule childRule = getAletheRule(childPf->getArguments()[0]);
-            Assert(
-                (childRule == AletheRule::CONTRACTION
-                 || childRule == AletheRule::REORDERING)
-                && getAletheRule(childPf->getChildren()[0]->getArguments()[0])
-                       == AletheRule::OR);
-            // get great grand child
-            std::shared_ptr<ProofNode> greatGrandChildPf =
-                childPf->getChildren()[0]->getChildren()[0];
-            Node greatGrandChildConclusion = greatGrandChildPf->getResult();
-            // Note that we need to add this proof node explicitly to cdp
-            // because cdp does not have a step for it. Rather it is only
-            // present in cdp as a descendant of childPf (which is in cdp), so
-            // if childPf is to be lost, then so will greatGrandChildPf.
-            addAletheStep(AletheRule::ASSUME,
-                          greatGrandChildConclusion,
-                          greatGrandChildConclusion,
-                          {},
-                          {},
-                          *cdp);
-            // equate it to what we expect, use equiv elim and resolution to
-            // obtain a proof the expected
-            Node expectedConclusion = childPf->getResult();
-            Assert(expectedConclusion == children[i]);
-            Node equiv = greatGrandChildConclusion.eqNode(expectedConclusion);
-            addAletheStep(AletheRule::ALL_SIMPLIFY,
-                          equiv,
-                          nm->mkNode(kind::SEXPR, d_cl, equiv),
-                          {},
-                          {},
-                          *cdp);
-            Node equivElim = nm->mkNode(kind::SEXPR,
-                                        {d_cl,
-                                         equiv.notNode(),
-                                         greatGrandChildConclusion.notNode(),
-                                         expectedConclusion});
-            addAletheStep(
-                AletheRule::EQUIV_POS2, equivElim, equivElim, {}, {}, *cdp);
-            Node newChild = nm->mkNode(kind::SEXPR, d_cl, expectedConclusion);
-            addAletheStep(AletheRule::RESOLUTION,
-                          newChild,
-                          newChild,
-                          {equivElim, equiv, greatGrandChildConclusion},
-                          d_resPivots
-                              ? std::vector<Node>{equiv,
-                                                  d_false,
-                                                  greatGrandChildConclusion,
-                                                  d_false}
-                              : std::vector<Node>(),
-                          *cdp);
-            newChildren[i] = newChild;
-            Trace("alethe-proof")
-                << "Reverted handling as a clause for converting "
-                << greatGrandChildConclusion << " into " << expectedConclusion
-                << std::endl;
+            newChildren[i] = nm->mkNode(kind::SEXPR, d_cl, children[i]);
           }
         }
       }

--- a/src/proof/alethe/alethe_post_processor.cpp
+++ b/src/proof/alethe/alethe_post_processor.cpp
@@ -83,8 +83,8 @@ bool AletheProofPostprocessCallback::update(Node res,
                                             CDProof* cdp,
                                             bool& continueUpdate)
 {
-  Trace("alethe-proof") << "...Alethe pre-update " << res << " " << id
-                        << " " << children << " / " << args << std::endl;
+  Trace("alethe-proof") << "...Alethe pre-update " << res << " " << id << " "
+                        << children << " / " << args << std::endl;
 
   NodeManager* nm = NodeManager::currentNM();
   std::vector<Node> new_args = std::vector<Node>();

--- a/src/proof/alethe/alethe_post_processor.h
+++ b/src/proof/alethe/alethe_post_processor.h
@@ -140,11 +140,11 @@ class AletheProofPostprocessCallback : protected EnvObj,
                            CDProof& cdp);
 
   /** Test whether resolution premise is wrongly derived as a non-singleton
-   *  clause. Fix if needed.
+   * clause. Fix if needed.
    *
-   *  If the premise is used as a singleton but its proof concludes a
-   *  non-singleton clause, a new proof of its derivation as a singleton is
-   *  added to cdp.
+   * If the premise is used as a singleton but its proof concludes a
+   * non-singleton clause, a new proof of its derivation as a singleton is added
+   * to cdp.
    */
   bool maybeReplacePremiseProof(Node premise, CDProof* cdp);
 

--- a/src/proof/alethe/alethe_post_processor.h
+++ b/src/proof/alethe/alethe_post_processor.h
@@ -139,6 +139,15 @@ class AletheProofPostprocessCallback : protected EnvObj,
                            const std::vector<Node>& args,
                            CDProof& cdp);
 
+  /** Test whether resolution premise is wrongly derived as a non-singleton
+   *  clause. Fix if needed.
+   *
+   *  If the premise is used as a singleton but its proof concludes a
+   *  non-singleton clause, a new proof of its derivation as a singleton is
+   *  added to cdp.
+   */
+  bool maybeReplacePremiseProof(Node premise, CDProof* cdp);
+
   /** Nodes corresponding to the Boolean values. */
   Node d_true;
   Node d_false;

--- a/test/regress/cli/CMakeLists.txt
+++ b/test/regress/cli/CMakeLists.txt
@@ -1100,6 +1100,7 @@ set(regress_0_tests
   regress0/proofs/issue9393-optResReconstruction-alethebug.smt2
   regress0/proofs/issue9515-alethe-skolems-crash.smt2
   regress0/proofs/issue9516-alethe-skolems-undef-in-proof.smt2
+  regress0/proofs/issue9531-alethe-resolution.smt2
   regress0/proofs/lfsc-test-1.smt2
   regress0/proofs/nomerge-alethe-pf.smt2
   regress0/proofs/open-pf-datatypes.smt2
@@ -2262,7 +2263,7 @@ set(regress_1_tests
   regress1/nl/proj-issue294.smt2
   regress1/nl/proj-issue297.smt2
   regress1/nl/proj-issue302.smt2
-  regress1/nl/proj-issue530.smt2 
+  regress1/nl/proj-issue530.smt2
   regress1/nl/quant-nl.smt2
   regress1/nl/red-exp.smt2
   regress1/nl/rewriting-sums.smt2

--- a/test/regress/cli/regress0/proofs/issue9531-alethe-resolution.smt2
+++ b/test/regress/cli/regress0/proofs/issue9531-alethe-resolution.smt2
@@ -1,0 +1,16 @@
+; COMMAND-LINE: --dump-proofs --proof-format-mode=alethe --dag-thresh=0 --proof-granularity=theory-rewrite
+; EXIT: 0
+; SCRUBBER: grep -v -E '.*'
+(set-logic QF_UF)
+(declare-const x2 Bool)
+(declare-const x Bool)
+(declare-sort I 0)
+(declare-fun u () I)
+(declare-fun o (I I) I)
+(declare-fun e () I)
+(assert (or x x2))
+(assert (or (= u (o u u)) (and (= u (o u u)) (or (not (= (o u (o u u)) (o (o (o e e) (o u (o u u))) (o e e)))) (= (o u (o u u)) (o (o e e) (o (o e e) (o u (o u u)))))))))
+(assert (= (o e e) (o u u)))
+(assert (or (= (o u (o u u)) (o (o e e) (o (o e e) (o u (o u u))))) (not (= (o u (o u u)) (o (o (o e e) (o u (o u u))) (o e e))))))
+(assert (not (= (o u (o u u)) (o (o u u) (o (o e e) (o u (o u u)))))))
+(check-sat)

